### ISSUE TITLE
Not handling case where pkg.bin is a string

### ISSUE THREAD

**Issue by jonas-lauber** [nicojs/node-link-parent-bin#4]
Hi, thanks for the great work. Your script has been really handy to me!
While I used it, I noticed that in a few cases it was generating scripts with strange names and I went through the code and figured out why. It turns out, that according to https://docs.npmjs.com/files/package.json#bin, one is allowed to use a string for the property **bin** of the **package.json**. In this case the value of _bin_ represents the path to the script and the property _name_ is used as the script name. So I think you script could easily be modified to cover this case by adding something like:

```
if (typeof pkg.bin === 'string') {
  return Promise.all(childPackages.map(childPackage =>
    this.linkBin(pkg.name, path.resolve(moduleDir, pkg.bin), childPackage)
    .catch(err => this.log.error(`Could not link bin ${pkg.bin} for child ${childPackage}.`, err))));
}
```

What do you think ?

**Comment by nicojs** [comment]
Awesome to hear this package is helpful to you. 

I wasn't aware of the string format. Your suggestion sounds great, are you willing to make a PR?

**Comment by jonas-lauber** [comment]
Hi,
I would love to, but can't manage to get the tests working. I run into all kinds of problems... But I must confess that I am quiet new to frontend development, so I am probably doing something wrong.

Here is my change: https://github.com/jonas-lauber/node-link-parent-bin/commit/340c17358de6fa54ab7572b01c7f6c583529a1f3#diff-bfc0d76123765cfb1e128b3f094b23de

I really feel bad I can't run the tests and do a proper pull request. That doesn't make much sense to fork the repo and do my stuff.... :-( Everybody could benefit from these fixes.

I'll give it another try if I find the time. (I just started a new project, so I am quiet busy right now..)

**Comment by nicojs** [comment]
> I would love to, but can't manage to get the tests working. I run into all kinds of problems... But I must confess that I am quiet new to frontend development, so I am probably doing something wrong.

That's alright, we've all been there ✌️ 

Let's start with the errors you face. You should be able to do this without errors:

```bash
git clone git@github.com:nicojs/node-link-parent-bin.git
cd node-link-parent-bin
npm i
npm t
```

The `npm t` command (short for `npm test`) should first run the unit tests. After that it runs the integration tests. 

Does that work at least?

> I really feel bad I can't run the tests and do a proper pull request. That doesn't make much sense to fork the repo and do my stuff.... :-( Everybody could benefit from these fixes.

Most definitely. However, doing your work in a fork is the way to go. You first need to fix stuff in your fork before you can create a Pull request. Let's take this one thing at a time.

**Comment by jonas-lauber** [comment]
Yes, this is working. I was also able to configure my IDE to run the tests and debug in TS. (This is however a bit shaky as I could't run a single test only... I had to run them all at once every times...)
So I added my unit tests, but now I have problems with the integration ones.  Because they have to run in a shell to access **npm**, I couldn't run them from my IDE directly. I wanted to reproduce the steps manually through a terminal, but couldn't either... When I try to run **"npm i"** from **sample**, I get a "install-local: command not found", which is true, as it is exactly the kind of problem your script is supposed to solve ^^

**Comment by nicojs** [comment]
> (This is however a bit shaky as I could't run a single test only... I had to run them all at once every times...)

The way to run a single test is by replacing `it(...)` with `it.only(...)`. It is not ideal (don't check in an `it.only` please), but you get used to it :blush: 

>  Because they have to run in a shell to access npm, I couldn't run them from my IDE directly

What IDE are you using? I'm using [vscode](https://code.visualstudio.com/) and the configuration is also added for debugging integration and unit tests. It runs on any device so you could try it as well. Just make sure you pull the latest sources (I needed to add a bit of config to make it work in the latest vscode version). Enable continuous compilation with `F1` "Run build task". Debug unit tests and integration tests from the debug pane on the left.

**Comment by jonas-lauber** [comment]
Ah ok, with Karma/Jasmine this was "fit" to run a single test ^^ Not really better ^^
I am using IntelliJ. Ok.. For any reason (maybe has to do with my npm version) the "pretest" wasn't run when doing a "npm run test" and thus the project not re-builded. So I was working with phantom code and that was pretty confusing ^^

But now my new tests are running fine. However I get:

> 1) program parse should parse [] to default values:
     TypeError: Path must be a string. Received undefined

My change have nothing to do with **program**... 

Unfortunately I don't have much more time to investigate this problem. So I'll let it be for now and if I have time in the next days go back to it. My quick analyse it that it has to do with sharing the commander instance with Mocha or not...

If mocha uses the same commander's version, the field _name is set by mocha when parsing the test command. And using the same version, it will be shared by npm into the root npm_modules and working with a single instance. If however the version is different, another instance is used in the test and the test fails with "Path must be a string. Received undefined" as the first is used to try to determine the name. 
This is done in **node_module/commander/index.js**. In _Command.prototype.parse_ we have:

> this._name = this._name || basename(argv[1], '.js');

So if the _name wasn't previously set, it tries to figure it out from argv and if this one is empty path.basename fails saying it is expecting a string.

**Comment by jonas-lauber** [comment]
(I just did a npm ls and this is exactly what happens... It fails when mocha uses its own version (i.e. instance) of commander...)

**Comment by nicojs** [comment]
Fixed since #6 (thanks @jan-molak)

Released in v0.2

**Comment by nicojs** [comment]
> If mocha uses the same commander's version, the field _name is set by mocha when parsing the test command. And using the same version, it will be shared by npm into the root npm_modules and working with a single instance.

Yes, I've noticed this about `commander`. Using the global `require('commander')` method is a bad practice and can even be labeled as a bug. This is my bad, I should have noticed it earlier. I also should have replied earlier (I've been busy 😢 ).

Commander has a nice little hidden feature: with `const command = new require('commander').Command()` you can create a new, independent command object, without state being shared across node_modules. I've created [commander.js issue 790](https://github.com/tj/commander.js/issues/790) to promote this behavior.

Thanks for all the help! And I'm sorry you ran into trouble.
